File: javascript/Exercise2/CodeToRefactor.js
class People {
    static Under16 = new Date(Date.now() - (15 * 365 * 24 * 60 * 60 * 1000));
    
    constructor(name, dob = null) {
        this.name = name;
        this.dob = dob || People.Under16;
    }
    
    getName() {
        return this.name;
    }
    
    getDOB() {
        return this.dob;
    }
}

class BirthingUnit {
    constructor() {
        // MaxItemsToRetrieve
        this._people = [];
    }

    /**
     * GetPeoples
     * @param {number} j
     * @returns {Array<Object>}
     */
    getPeople(i) {
        for (let j = 0; j < i; j++) {
            try {
                // Creates a dandon Name
                let name = "";
                const random = Math.random();
                if (Math.floor(random * 2) === 0) {
                    name = "Bob";
                } else {
                    name = "Betty";
                }
                // Adds new people to the list
                const age = Math.floor(Math.random() * (85 - 18) + 18) * 356;
                const birthDate = new Date(Date.now() - (age * 24 * 60 * 60 * 1000));
                this._people.push(new People(name, birthDate));
            } catch (e) {
                // Dont think this should ever happen
                throw new Error("Something failed in user creation");
            }
        }
        return this._people;
    }

    getBobs(olderThan30) {
        const thirtyYearsAgo = new Date(Date.now() - (30 * 356 * 24 * 60 * 60 * 1000));
        return olderThan30 
            ? this._people.filter(x => x.name === "Bob" && x.dob >= thirtyYearsAgo)
            : this._people.filter(x => x.name === "Bob");
    }

    getMarried(p, lastName) {
        if (lastName.includes("test"))
            return p.name;
        if ((p.name.length + lastName).length > 255) {
            (p.name + " " + lastName).substring(0, 255);
        }

        return p.name + " " + lastName;
    }
}

module.exports = { People, BirthingUnit };

File: ruby/Exercise2/CodeToRefactor.rb
require 'date'

module Refactor
  class People
    UNDER_16 = DateTime.now - (15 * 365)
    
    attr_reader :name, :dob
    
    def initialize(name, dob = nil)
      @name = name
      @dob = dob || UNDER_16
    end
  end

  class BirthingUnit
    # MaxItemsToRetrieve
    def initialize
      @people = []
    end

    # GetPeoples
    # @param j
    # @return Array<Object>
    def get_people(i)
      i.times do |j|
        begin
          # Creates a dandon Name
          name = ""
          if rand(0..1) == 0
            name = "Bob"
          else
            name = "Betty"
          end
          # Adds new people to the list
          @people << People.new(name, DateTime.now - (rand(18..85) * 356))
        rescue => e
          # Dont think this should ever happen
          raise "Something failed in user creation"
        end
      end
      @people
    end

    private

    def get_bobs(older_than_30)
      if older_than_30
        @people.select { |x| x.name == "Bob" && x.dob >= (DateTime.now - (30 * 356)) }
      else
        @people.select { |x| x.name == "Bob" }
      end
    end

    public

    def get_married(p, last_name)
      return p.name if last_name.include?("test")
      
      if (p.name.length + last_name).length > 255
        (p.name + " " + last_name)[0, 255]
      end

      p.name + " " + last_name
    end
  end
end
